release.sh

```diff
@@ -1,5 +1,8 @@
+#/usr/bin/env bash
+set -e
 
-
+# Install `twine` through `pip` if it hasn't been installed yet
+command -v twine >/dev/null 2>&1 || pip install twine
```

Comment:
This is OK.  My preferred method is to have `twine` and everything else development related on a `dev` hook inside `setup.py` so that everything is installed from one place (and it's clear where to set version numbers if needed).  But it's fine if you prefer to have it like this.

Reply:
twine is the only package that was still required for the `dev` extra, which then is really only necessary for releasing. Think this solution makes more sense then, even though typically you will have it available anyway in your release or dev environment. The reason I split `pre-commit` and `tests` is that often we want to run the `tests` for a range of Python versions. But the pre-commit we only have to run for one (usually latest) and often those dependencies might no longer support the oldest Python version. This forces us either to downgrade those pre-commit dependencies or simply split the extras. I think the latter is the cleaner solution.